### PR TITLE
[DeepSeek v4] Fixes MTP draft proposal sampling, 50% boost when C=1

### PR DESCRIPTION
**Pull request by jasl**
## Summary

This PR fixes MTP draft proposal sampling so the rejection sampler receives the correct draft-token probability distribution instead of treating MTP draft tokens as probability-less proposals.

Before this change, MTP draft proposals did not preserve the proposal probability distribution needed by rejection sampling: `GPUModelRunner` passed `draft_probs=None`, and the draft sampling helper only partially handled non-greedy sampling parameters. That is acceptable for purely greedy draft proposals, but it is wrong for random sampling and for sampling parameters such as temperature, top-k, and top-p. It can also misalign draft probabilities after request batch reordering.

This PR:

- Computes draft probabilities when MTP/EAGLE proposals use non-greedy sampling.
- Applies temperature, top-k, top-p, and per-request generators to draft proposal sampling.
- Stores per-position draft probabilities from the proposer.
- Packs draft probabilities for rejection sampling according to the previous batch positions.
- Handles reordered batches and requests with different accepted draft-token counts.
- Adds focused MTP tests for draft probability recording, sampling filters, parallel drafting, and batch reorder packing.

## Before / After

| Area | Before | After |
|---|---|---|
| Draft proposal probabilities | MTP did not preserve the proposal probability distribution for rejection sampling. | MTP records draft proposal probabilities when they are needed. |
| Rejection sampler input | `GPUModelRunner` passed `draft_probs=None`. | `GPUModelRunner` passes packed draft probabilities when available. |
| Sampling params | Draft proposal sampling only partially handled non-greedy sampling parameters. | Draft proposal probabilities apply temperature, top-k, top-p, and per-request generators. |
| Multi-token MTP | No retained `[batch, num_spec_tokens, vocab]` draft probability tensor. | Proposer records per-position draft probabilities for multi-token MTP and parallel drafting. |
| Batch reorder | Draft probabilities could be read in current batch order after requests were reordered. | Draft probabilities are packed through `prev_positions`, preserving request alignment. |
| Test coverage | Existing MTP test mostly checked proposal shape/model wiring. | Tests cover draft probs, top-k filtering, parallel drafting, rejection-packing, and reordered batches. |

## Downstream Validation Context

This PR is intentionally hardware-agnostic, but it was validated as part of the DeepSeek V4 MTP bring-up work where incorrect or missing draft probabilities made it hard to reason about MTP behavior.

On the SM120 DeepSeek V4 evaluation branch, this generic fix is one prerequisite for the MTP path to become usable. With this fix plus separate SM120 sparse MLA runtime optimizations, MTP becomes a clear positive path versus no-MTP:

| Concurrency | no-MTP output tok/s | MTP output tok/s | MTP delta | no-MTP TPOT | MTP TPOT | MTP acceptance |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 97.36 | 150.86 | +54.94% | 9.97 ms | 6.35 ms | 57.50% |
| 4 | 268.61 | 305.78 | +13.84% | 14.00 ms | 12.51 ms | 55.60% |
| 8 | 418.22 | 528.05 | +26.26% | 17.41 ms | 14.27 ms | 57.18% |

Notes:

- The throughput numbers above include additional SM120-specific runtime work and are not claimed as the isolated effect of this PR.
- This PR fixes the generic MTP probability correctness issue that those runtime tests depend on.
- The acceptance rate staying around 55-57% after this fix indicates MTP is now using a coherent draft-probability path rather than a missing-probability rejection path.

### Proposer

`SpecDecodeBaseProposer` now has a `_sample_draft_token_ids` helper:

- Greedy requests continue to use the existing greedy path.
- Non-greedy requests compute logits, probabilities, and sampled draft tokens.
- Draft probabilities are stored as `self.draft_probs`.

For multi-token MTP, probabilities are stacked into:

```text
[batch_size, num_speculative_tokens, vocab_size]
```

For parallel drafting, the same shape is produced directly from the parallel logits.

### Draft Sampling

`compute_probs_and_sample_next_token` now:

- Expands per-request sampling tensors to match draft-logit rows.
- Applies temperature.
- Applies top-k/top-p filtering.
- Preserves generator mapping when a request expands into multiple draft positions.
- Returns both sampled token ids and the probability tensor needed by rejection sampling.

### GPUModelRunner

`GPUModelRunner` now stores `_draft_probs` after proposing draft tokens and passes packed draft probabilities into `rejection_sampler`.

The packing logic handles:

- Flat `[total_draft_tokens, vocab]` tensors.
- Structured `[batch, num_spec_tokens, vocab]` tensors.
- Requests with fewer valid draft tokens than `num_spec_tokens`.
- Batch reorder via `prev_positions`.

## Test Plan

Run on the branch:

```bash
python -m pytest -q tests/v1/spec_decode/test_mtp.py
```

Verified result:

```text
7 passed, 16 warnings
```

## Scope

This PR is a generic MTP correctness fix. It does not include the SM120 sparse MLA runtime optimizations, DeepSeek V4 SM120 integration work, or benchmark-only evaluation branch changes.

### CODE REVIEW

**Comment by jasl** [comment]
@johnnynunez Please help to take a look

**Comment by jasl** [comment]
image

the final benchmark from my own branch (2 x RTX Pro 6000)

**Comment by ywang96** [comment]
@jasl Very much appreciate the PR! We're taking a look at this!

**Comment by TheEpicDolphin** [comment]
How many speculative tokens were used for the benchmarks? If more than 1, would you mind sharing the per-position acceptance rates?

**Comment by jasl** [comment]
> How many speculative tokens were used for the benchmarks? If more than 1, would you mind sharing the per-position acceptance rates?

We used `num_speculative_tokens=2`.

The benchmark was run on the SM120 DeepSeek V4 evaluation branch with TP=2, PP=1, fp8 KV cache, `max_model_len=8192`, `max_num_seqs=24`, random 1024 input / 1024 output, 48 prompts, and concurrency 1/4/8. The serve command used:

`--speculative-config '{"method":"mtp","num_speculative_tokens":2}'`

Per-position acceptance rates from those runs:

| Concurrency | Overall acceptance | Position 0 | Position 1 |
|---:|---:|---:|---:|
| 1 | 57.50% | 76.94% | 38.07% |
| 4 | 55.60% | 75.70% | 35.50% |
| 8 | 57.18% | 78.60% | 35.76% |

**Comment by benchislett** [comment]
Is this a duplicate of https://github.com/vllm-project/vllm/pull/40269?

**Comment by benchislett** [comment]
@jasl we have not always seen improvements to acceptance rate when using this feature (it is already available when using ModelRunnerV2). Can you please compare MTP-before to MTP-after, instead of no-MTP to MTP-after?

I'm not actually sure this is a bug. What exactly happens when you try to run MTP on SM120 without this PR's change?

**Comment by jasl** [comment]
> @jasl we have not always seen improvements to acceptance rate when using this feature (it is already available when using ModelRunnerV2). Can you please compare MTP-before to MTP-after, instead of no-MTP to MTP-after?
> 
> I'm not actually sure this is a bug. What exactly happens when you try to run MTP on SM120 without this PR's change?

Without the fix, I can see a huge degradation when enabling MTP, and I confirmed the same degradation on GH200 (but I don't have the data, it's my friend who found it)

**Comment by jasl** [comment]
I just found a chat history, he said: spec bench 2nd token less than 50%, unless you use big EP

**Comment by jasl** [comment]
> Is this a duplicate of #40269?

GPT helps me summarize:

#40269 wires draft_probs through the probabilistic draft_model/EAGLE rejection path. This PR focuses on the MTP path: MTP draft proposals need to sample from the same non-greedy distribution used for proposal probabilities, including temperature, top-k/top-p, per-request generators, multi-token/parallel drafting probability storage, and packing through prev_positions after batch reordering.

There is some shared plumbing because both fixes eventually pass draft_probs into RejectionSampler. If #40269 lands first, I’m happy to rebase on top of it and reduce this PR to the MTP-specific delta.

**Comment by jasl** [comment]
Another GPT summarizes (sorry, I can't speak without him 😄 )

I agree that MTP itself is not guaranteed to improve throughput; the gain depends heavily on the model, acceptance rate, and serving shape. The issue I am trying to fix here is narrower: in the V1 MTP path, non-greedy draft proposals do not preserve/pass the proposal probabilities needed by rejection sampling. That is independent of SM120.

The practical symptom is that on DeepSeek V4 with temperature=1, enabling MTP can become a large regression before this fix. I also heard similar degradation on GH200, which suggests this is not SM120-specific. Deployment shape still matters: with low EP or poor second-token acceptance, MTP may remain marginal even after the probability-path fix. But this PR fixes the probability correctness issue so that any remaining gain/loss reflects the actual MTP/model/system tradeoff rather than missing or misaligned draft probabilities.

**Comment by jasl** [comment]
@benchislett 
I did a test on B200 * 2 (CUDA 13.0) with official vLLM 0.20.0, I can see a slight improvement on C=1, but unstable on C=4 and C=8

Setup: 2x NVIDIA B200, TP=2, DeepSeek-V4-Flash, `kv-cache-dtype=fp8`, `max-model-len=8192`, `random-input-len=1024`, `random-output-len=512`, `num-prompts=32`. MTP used `{"method":"mtp","num_speculative_tokens":2}`.

| Concurrency | no-MTP success/fail | no-MTP out tok/s | no-MTP TPOT | MTP success/fail | MTP out tok/s | MTP TPOT | MTP acceptance | Notes |
|---:|---:|---:|---:|---:|---:|---:|---:|---|
| 1 | 32 / 0 | 123.29 | 7.93 ms | 32 / 0 | 141.90 | 6.76 ms | 22.65% | MTP improved throughput by ~15.1%. |
| 4 | 32 / 0 | 370.26 | 9.39 ms | 4 / 28 | 0.02 | 6877.20 ms | N/A | MTP path hung, then EngineCore died. |
| 8 | 32 / 0 | 510.63 | 11.21 ms | 8 / 24 | 9.92 | 9.78 ms | 28.12% | Fresh MTP server start, still ended with EngineDeadError. |

**Comment by benchislett** [comment]
I'm closing this PR. If you have indeed confirmed a large MTP acceptance rate degradation, please file a github issue with instructions to reproduce.

**Comment by jasl** [comment]
> I'm closing this PR. If you have indeed confirmed a large MTP acceptance rate degradation, please file a github issue with instructions to reproduce.

I've provided the charts above, and in my 2*B200 GPU, the MTP can't even work, what kind of instructions to reproduce do you want? Have you tried this configuration?

**Comment by benchislett** [comment]
the chart above still seems to compare no-MTP to MTP. This is not answering my question about MTP with prob sampling v.s. MTP with greedy sampling

**Comment by benchislett** [comment]
there is no included acceptance rate information about the existing MTP to compare against

**Comment by benchislett** [comment]
and it is not clear to me how any of this is related to the crashing issue. that seems like a completely separate problem

**Comment by jasl** [comment]
I reran the comparison requested here on a fresh 4x B200 host using vLLM v0.20.0 as the common baseline. To isolate this PR, I kept the same installed v0.20.0 native wheel/CUDA runtime and cherry-picked only the three Python MTP commits from this PR onto the v0.20.0 source files.

This compares existing v0.20.0 MTP greedy draft proposal vs this PR’s MTP probability draft proposal, not no-MTP vs MTP.

Setup:
- Model: deepseek-ai/DeepSeek-V4-Flash
- TP=4, EP enabled
- speculative_config: {"method":"mtp","num_speculative_tokens":2}
- random dataset: input_len=1024, output_len=256
- num_prompts=32, max_concurrency=1
- temperature=1.0, top_p=1.0, ignore_eos=true

Results:

| Variant | Output tok/s | Mean TPOT | Acceptance | Mean acceptance length | Accepted / Drafted | Pos0 / Pos1 acceptance |
|---|---:|---:|---:|---:|---:|---:|
| v0.20.0 existing MTP greedy proposal | 113.40 | 7.98 ms | 10.28% | 1.21 | 1393 / 13548 | 14.90% / 5.67% |
| This PR, MTP probability proposal | 168.71 | 5.15 ms | 49.38% | 1.99 | 4062 / 8226 | 70.02% / 28.74% |

So under the same MTP configuration, this PR improves output throughput by about 48.8%, reduces mean TPOT by about 35.4%, and raises draft acceptance from 10.28% to 49.38%.

I agree the crash mentioned elsewhere is a separate issue and should not be used as evidence for this PR. This PR is specifically about fixing the non-greedy MTP proposal/probability path so MTP behavior reflects the intended sampling distribution.

**Comment by jasl** [comment]
@benchislett, please have a look

**Comment by benchislett** [comment]
Thanks, this is useful. But you are benchmarking with random data. This is the ideal case for probabilistic sampling and not reflective of real workloads. Can you run MT-Bench or similar?

For example: `vllm bench serve --dataset-name hf --dataset-path philschmid/mt-bench --num-prompts 80 --max-concurrency 1 --model $MODEL --base-url $BASE_URL --temperature 1.0`

note also that when running with lower temperature (e.g. 0.0) the gap would be lower

**Comment by jasl** [comment]
Will do.

BTW, when I'm working on the SM120 branch, I see that Deepseek v4 at temperature=0 shows responses unstable.
Official recommends temperature=1 and top_k=1 work great.

**Comment by benchislett** [comment]
Yes, for DSV4 temp=1 is recommended. Just mentioning temp=0 since it is often used when measuring specdec acceptance rates, hence why the probabilistic feature has not been as high priority in the past

**Comment by benchislett** [comment]
We will move forward with #40269 for now

**Comment by jasl** [comment]
Thanks, this is a fair point. I reran the comparison on a fresh 4x B200 host with vLLM v0.20.0 native extensions as the common runtime. For the patched variant, I cherry-picked only this PR's Python MTP changes on top of v0.20.0, so the native/CUDA runtime was identical.

I tried the MT-Bench command shape with `philschmid/mt-bench`, but DeepSeek-V4-Flash's tokenizer in this setup does not define `chat_template`, so the benchmark errors unless `--skip-chat-template` is added. The numbers below are therefore MT-Bench first-turn prompts used directly as completion prompts, not chat-template-rendered chat requests.

Setup: DeepSeek-V4-Flash, TP=4 + EP, MTP `num_speculative_tokens=2`, max concurrency 1, temperature 1.0, 80 prompts.

| Variant | Output tok/s | Mean TPOT | Acceptance | Accept len | Accepted / Drafted | Pos0 / Pos1 |
|---|---:|---:|---:|---:|---:|---:|
| v0.20.0 baseline MTP | 203.48 | 4.64 ms | 54.82% | 2.10 | 10320 / 18826 | 70.83% / 38.81% |
| This PR | 195.67 | 4.75 ms | 65.28% | 2.31 | 10960 / 16790 | 82.80% / 47.75% |

This supports your concern about the random benchmark: random prompts made the baseline greedy MTP path collapse to about 10% acceptance, so that benchmark overstated the throughput benefit. On MT-Bench prompts, baseline MTP already has much higher acceptance.

The PR still improves the probability/proposal behavior materially: acceptance rate increases by +10.46 pp, acceptance length improves from 2.10 to 2.31, and it accepts more tokens with fewer drafted tokens. However, this MT-Bench run does not show a throughput win. I will avoid claiming a universal throughput improvement and frame this PR as fixing non-greedy MTP draft probability correctness; throughput impact depends on workload and temperature.

I also agree the crash issue is separate from this MTP probability-path fix and should be discussed separately.

**Comment by benchislett** [comment]
Okay. Please file a github issue for the crash problem. For probabilistic sampling refer to #40269
